File: 2024/clojure/src/day01.clj
(ns day01
  (:require
   [clojure.string :as str]))

(defn parse-input [file-path]
  (let [lines (str/split-lines (slurp file-path))]
    (map (fn [line]
           (let [[x y] (str/split line #"\s+")]
             [(Integer/parseInt x) (Integer/parseInt y)]))
         lines)))

(defn transpose [pairs]
  (apply map vector pairs))

(defn sort-transposed [transposed]
  (map sort transposed))

(defn zip-and-difference [transposed]
  (map (fn [[x y]] (abs (- y x))) (apply map vector transposed)))

(defn sum [numbers]
  (reduce + numbers))

(defn frequency-map [numbers]
  (reduce (fn [freq-map item]
            (update freq-map item (fnil inc 0)))
          {}
          numbers))

(defn multiply-by-frequency [transposed]
  (let [[list1 list2] transposed
        freq-map (frequency-map list2)]
    (map (fn [x] (* x (get freq-map x 0))) list1)))

(def parsed (parse-input "data/day01.txt"))
(def transposed (transpose parsed))
(def sorted-transposed (sort-transposed transposed))
(def differences (zip-and-difference sorted-transposed))

(let [answer (sum differences)]
  (println "Part 1:" answer))

(let [results (multiply-by-frequency transposed)
      answer (sum results)]
  (println "Part 2:" answer))

File: 2024/python/day01.py
import argparse
import re

import numpy as np

from utils import read_lines

parser = argparse.ArgumentParser()
parser.add_argument("input", type=str, help="Path to input file")


def parse_line(line: str) -> list[int]:
    return [int(part) for part in re.split(r"\s+", line)]


def part_1(x):
    x = np.sort(x, axis=0)
    x = np.diff(x, axis=1)
    x = np.abs(x)
    return np.sum(x)


def part_2(x):
    x = np.transpose(x)
    labels, counts = np.unique_counts(x[1])
    label_to_count = {label: count for label, count in zip(labels, counts)}
    result = np.array([v * label_to_count.get(v, 0) for v in x[0]])
    return np.sum(result)


def main():
    args = parser.parse_args()
    x = np.array(read_lines(args.input, cast=parse_line))

    print("Part 1:", part_1(x))
    print("Part 2:", part_2(x))


if __name__ == "__main__":
    main()
